File: Algorithms/844.比较含退格的字符串/backspaceCompare.py
# 模拟编辑，时间复杂度O(|S|+|T|)，空间复杂度O(|S|+|T|)
class Solution:
    def backspaceCompare(self, S: str, T: str) -> bool:
        return edit(S) == edit(T)
    
def edit(s):
    result = []
    for c in s:
        if c == '#':
            if result:
                result.pop()
        else:
            result.append(c)
    return ''.join(result)

File: Algorithms/844.比较含退格的字符串/backspaceCompare.java
// 官方题解：双指针法，时间复杂度O(|S|+|T|)，空间复杂度O(1)
class Solution {
    public boolean backspaceCompare(String S, String T) {
        int i = S.length() - 1, j = T.length() - 1;
        int skipS = 0, skipT = 0;

        while (i >= 0 || j >= 0) {
            while (i >= 0) {
                if (S.charAt(i) == '#') {
                    skipS++;
                    i--;
                }
                else if (skipS > 0) {
                    skipS--;
                    i--;
                }
                else
                    break;
            }
            while (j >= 0) {
                if (T.charAt(j) == '#') {
                    skipT++;
                    j--;
                }
                else if (skipT > 0) {
                    skipT--;
                    j--;
                }
                else
                    break;
            }
            if (i >= 0 && j >= 0) {
                if (S.charAt(i) != T.charAt(j))
                    return false;
            }
            else {
                if (i >= 0 || j >= 0)
                    return false;
            }
            i--;
            j--;
        }
        return true;
    }
}
